File: lab/workflow.py
"""
functions for setting the workflow of a repository
"""

# SPDX-FileCopyrightText: 2020 Jonah Brüchert <jbb@kaidan.im>
#
# SPDX-License-Identifier: GPL-2.0-or-later

import argparse

from lab.config import RepositoryConfig, Workflow


def parser(
    subparsers: argparse._SubParsersAction,  # pylint: disable=protected-access
) -> argparse.ArgumentParser:
    """
    Subparser for workflow command
    :param subparsers: subparsers object from global parser
    :return: merge request creation subparser
    """
    workflow_parser: argparse.ArgumentParser = subparsers.add_parser(
        "workflow", help="Set the workflow to use for a project"
    )

    group = workflow_parser.add_mutually_exclusive_group(required=True)
    group.add_argument(
        "--fork",
        help="Set the fork workflow (branch in a fork of the upstream repository)",
        action="store_true",
    )
    group.add_argument(
        "--workbranch",
        help="Set the work branch workflow (branch in the upstream repository)",
        action="store_true",
    )
    return workflow_parser


def run(args: argparse.Namespace) -> None:
    """
    run workflow command
    :param args: parsed arguments
    """
    repository_config = RepositoryConfig()
    if args.fork:
        repository_config.set_workflow(Workflow.FORK)
    elif args.workbranch:
        repository_config.set_workflow(Workflow.WORKBRANCH)
    repository_config.save()

File: src/workflow.cpp
#include "workflow.h"

#include <pybind11/embed.h>

namespace py = pybind11;

void Workflow::run(bool fork, bool workbranch)
{
    auto repository_config = py::module::import("lab.config").attr("RepositoryConfig")();
    auto workflowEnum = py::module::import("lab.config").attr("Workflow");
    if (fork) {
        repository_config.attr("set_workflow")(workflowEnum(int(WorkflowType::ForkWorkflow)));
    } else if (workbranch) {
        repository_config.attr("set_workflow")(workflowEnum(int(WorkflowType::WorbranchWorkflow)));
    }
    repository_config.attr("save")();
}
